File: locus-api-android/src/main/java/locus/api/android/features/computeTrack/ComputeTrackParameters.java
package locus.api.android.features.computeTrack;

import java.io.IOException;

import locus.api.objects.Storable;
import locus.api.objects.extra.ExtraData;
import locus.api.objects.extra.Location;
import locus.api.utils.DataReaderBigEndian;
import locus.api.utils.DataWriterBigEndian;

/**
 * Created by menion on 10. 7. 2014.
 * Class is part of Locus project
 */
public class ComputeTrackParameters extends Storable {

    // type of track that should be compute
    private int mType;
    // <code>True</code> if user wants with navigation orders
    private boolean mComputeInstructions;
    // flag if parameters has defined starting direction
    private boolean mHasDirection;
    // direction value
    private float mDirection;
    // list of locations (via-points)
    private Location[] mLocs;

    /**
     * Empty constructor for 'Storable' class. Do not use directly
     */
    @SuppressWarnings("unused")
    public ComputeTrackParameters() {}

    public ComputeTrackParameters(int type, Location[] locs) {
        super();

        // set parameters
        this.mType = type;
        if (locs == null || locs.length < 2) {
            throw new IllegalArgumentException("'locs' parameter" +
                    " cannot be 'null' or smaller then 2");
        }
        this.mLocs = locs;
    }

    public ComputeTrackParameters(byte[] data) throws IOException {
        super(data);
    }

    public int getType() {
        return mType;
    }

    public boolean isComputeInstructions() {
        return mComputeInstructions;
    }

    public void setComputeInstructions(boolean computeInstructions) {
        this.mComputeInstructions = computeInstructions;
    }

    // STARTING DIRECTION

    /**
     * Check if direction is currently defined.
     * @return <code>true</code> if direction for first point is set
     */
    public boolean hasDirection() {
        return mHasDirection;
    }

    /**
     * Get direction angle of movement at first point.
     * @return define direction angle
     */
    public float getCurrentDirection() {
        return mDirection;
    }

    /**
     * Set direction angle of movement at first point in degress.
     * @param direction direction angle
     */
    public void setCurrentDirection(float direction) {
        this.mDirection = direction;
    }

    // LOCATIONS

    /**
     * Get list of defined via-points (location).
     * @return list of locations
     */
    public Location[] getLocations() {
        return mLocs;
    }

    /**************************************************/
    // STORABLE PART
    /**************************************************/

    @Override
    protected int getVersion() {
        return 1;
    }

    @Override
    public void reset() {
        mType = ExtraData.VALUE_RTE_TYPE_CAR;
        mComputeInstructions = true;
        mHasDirection = false;
        mDirection = 0;
        mLocs = new Location[0];
    }

    @Override
    protected void readObject(int version, DataReaderBigEndian dr) throws IOException {
        mType = dr.readInt();
        mComputeInstructions = dr.readBoolean();
        mDirection = dr.readFloat();
        mLocs = new Location[dr.readInt()];
        for (int i = 0, m = mLocs.length; i < m; i++) {
            mLocs[i] = new Location(dr);
        }

        // V1

        if (version >= 1) {
            mHasDirection = dr.readBoolean();
        }
    }

    @Override
    protected void writeObject(DataWriterBigEndian dw) throws IOException {
        dw.writeInt(mType);
        dw.writeBoolean(mComputeInstructions);
        dw.writeFloat(mDirection);
        dw.writeInt(mLocs.length);
        for (Location mLoc : mLocs) {
            dw.writeStorable(mLoc);
        }

        // V1

        dw.writeBoolean(mHasDirection);
    }
}

File: locus-api-android/src/main/java/locus/api/android/features/computeTrack/ComputeTrackParameters.kt
/**
 * Created by menion on 10. 7. 2014.
 * Class is part of Locus project
 */
package locus.api.android.features.computeTrack

import locus.api.objects.Storable
import locus.api.objects.extra.GeoDataExtra
import locus.api.objects.extra.Location
import locus.api.utils.DataReaderBigEndian
import locus.api.utils.DataWriterBigEndian
import java.io.IOException

/**
 * Container for parameters needed to compute route over [ComputeTrackService] instance.
 *
 * Empty constructor for 'Storable' class. Do not use directly
 */
@Suppress("MemberVisibilityCanBePrivate")
class ComputeTrackParameters() : Storable() {

    /**
     * Type of track that should be compute.
     */
    var type: Int = GeoDataExtra.VALUE_RTE_TYPE_CAR
        private set
    /**
     * Flag if user wants with navigation orders.
     */
    var isComputeInstructions: Boolean = true
    /**
     * Direction angle of movement at first point [°].
     */
    var currentDirection: Float = 0.0f
        set(value) {
            field = value
            hasDirection = true
        }
    /**
     * Flag if parameters has defined starting direction.
     */
    var hasDirection: Boolean = false
        private set
    /**
     * List of defined via-points (location)
     */
    var locations: Array<Location> = arrayOf()
        private set

    /**
     * Create new parameters for track compute.
     *
     * @param type type of track to compute
     * @param locs list of defined points over which route should be computed
     */
    constructor(type: Int, locs: Array<Location>) : this() {
        this.type = type
        if (locs.size < 2) {
            throw IllegalArgumentException("'locs' parameter" + " cannot be 'null' or smaller then 2")
        }
        this.locations = locs
    }

    //*************************************************
    // STORABLE PART
    //*************************************************

    override fun getVersion(): Int {
        return 1
    }

    @Throws(IOException::class)
    override fun readObject(version: Int, dr: DataReaderBigEndian) {
        type = dr.readInt()
        isComputeInstructions = dr.readBoolean()
        currentDirection = dr.readFloat()

        // read locations
        val size = dr.readInt()
        val locs = arrayListOf<Location>()
        for (i in 0 until size) {
            locs.add(Location().apply { read(dr) })
        }
        locations = locs.toTypedArray()

        // V1
        if (version >= 1) {
            hasDirection = dr.readBoolean()
        }
    }

    @Throws(IOException::class)
    override fun writeObject(dw: DataWriterBigEndian) {
        dw.writeInt(type)
        dw.writeBoolean(isComputeInstructions)
        dw.writeFloat(currentDirection)

        // write locations
        dw.writeInt(locations.size)
        for (mLoc in locations) {
            dw.writeStorable(mLoc)
        }

        // V1
        dw.writeBoolean(hasDirection)
    }
}
